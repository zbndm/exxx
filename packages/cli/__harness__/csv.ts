const header =
  '"type","continent","iso_country","iso_region","iata_code"' + '\n';

const airportsByIATA = {
  cph: '"large_airport","EU","DK","DK-84","CPH"',
  iev: '"medium_airport","EU","UA","UA-32","IEV"',
  kbp: '"large_airport","EU","UA","UA-32","KBP"',
};

type RealAirportIATA = keyof typeof airportsByIATA;

function isListOfRealAirports(
  iataList: ReadonlyArray<string>
): iataList is ReadonlyArray<RealAirportIATA> {
  return iataList.every((item) => item in airportsByIATA);
}

export function airports(
  ...args: ReadonlyArray<ReadonlyArray<RealAirportIATA> | ReadonlyArray<string>>
) {
  const list = args.flatMap((dataItems) =>
    isListOfRealAirports(dataItems)
      ? dataItems.map((iata) => airportsByIATA[iata] + '\n')
      : dataItems.map((listItem) => listItem + '\n')
  );

  return [header, ...list].join('');
}
