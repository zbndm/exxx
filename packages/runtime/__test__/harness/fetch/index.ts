import { Readable } from 'stream';

export function stream(content: string) {
  const ref = {
    mock: null as unknown as jest.SpyInstance<
      Promise<import('node-fetch').Response>
    >,
    response: null as unknown as import('node-fetch').Response,
  };

  beforeEach(() => {
    const fetch = jest.requireActual<typeof import('node-fetch')>('node-fetch');

    const body = Readable.from([Buffer.from(content)]);
    ref.response = new fetch.Response(body);

    ref.mock = jest.spyOn(fetch, 'default').mockResolvedValue(ref.response);
  });

  afterEach(() => ref.mock.mockReset());

  afterAll(() => ref.mock.mockRestore());

  return ref;
}
