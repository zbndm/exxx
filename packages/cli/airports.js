/* eslint-disable camelcase */
import csv from 'csv-stream';
import stringify from 'fast-stable-stringify';
import fetch from 'node-fetch';
import prettier from 'prettier';

export function main() {
  const url = 'https://ourairports.com/data/airports.csv';

  const codes = {};
  function filterAirports(airport) {
    if (!airport.iata_code || !airport.type.endsWith('airport')) {
      return;
    }

    const { iata_code, iso_country, iso_region } = airport;

    const skipRules = {
      [`Skipping duplicate IAIA ${iata_code}`]:
        Object.prototype.hasOwnProperty.call(codes, iata_code),

      [`Skipping invalid IAIA ${iata_code}`]:
        !/[A-Z]+/.test(iata_code) || /[^A-Z0-9]+/.test(iata_code),

      [`Skipping IAIA ${iata_code} because country ${iso_country} is not in region ${iso_region}`]:
        iso_country.toUpperCase() === 'NA' &&
        iso_region.substring(0, 2).toUpperCase() === 'US',
    };

    for (const [reason, condition] of Object.entries(skipRules)) {
      if (condition) {
        process.stderr.write(reason + '\n');

        return;
      }
    }

    const item = {
      continent: /\d+/.test(airport.continent) ? undefined : airport.continent,
      iso_country,
      // type: airport.type,
      iso_region,
    };

    codes[iata_code] = item;
  }

  return new Promise((resolve, reject) => {
    fetch(url).then(
      (r) =>
        r.body
          .pipe(csv.createStream({ enclosedChar: '"' }))
          .on('data', filterAirports)
          .on('end', () => {
            const output = prettier.format(stringify(codes), {
              semi: false,
              parser: 'json',
            });

            resolve(output);
          })
          .on('error', reject),
      reject
    );
  });
}
