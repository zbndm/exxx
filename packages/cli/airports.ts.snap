// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Airports Script filtering country-specific should skip 1`] = `
"{ \\"CPH\\": { \\"continent\\": \\"EU\\", \\"iso_country\\": \\"DK\\", \\"iso_region\\": \\"DK-84\\" } }
"
`;

exports[`Airports Script filtering duplicate airports should skip 1`] = `
"{ \\"CPH\\": { \\"continent\\": \\"EU\\", \\"iso_country\\": \\"DK\\", \\"iso_region\\": \\"DK-84\\" } }
"
`;

exports[`Airports Script filtering invalid IATA should skip 1`] = `
"{ \\"CPH\\": { \\"continent\\": \\"EU\\", \\"iso_country\\": \\"DK\\", \\"iso_region\\": \\"DK-84\\" } }
"
`;

exports[`Airports Script filtering non-airports should skip 1`] = `
"{ \\"CPH\\": { \\"continent\\": \\"EU\\", \\"iso_country\\": \\"DK\\", \\"iso_region\\": \\"DK-84\\" } }
"
`;

exports[`Airports Script filtering with numeric continent should keep the record w/o continent 1`] = `
"{
  \\"CPH\\": { \\"continent\\": \\"EU\\", \\"iso_country\\": \\"DK\\", \\"iso_region\\": \\"DK-84\\" },
  \\"NUMERIC\\": { \\"iso_country\\": \\"US\\", \\"iso_region\\": \\"US-LA\\" }
}
"
`;

exports[`Airports Script output formatting should pretty-print 1`] = `
"{
  \\"CPH\\": { \\"continent\\": \\"EU\\", \\"iso_country\\": \\"DK\\", \\"iso_region\\": \\"DK-84\\" },
  \\"IEV\\": { \\"continent\\": \\"EU\\", \\"iso_country\\": \\"UA\\", \\"iso_region\\": \\"UA-32\\" },
  \\"KBP\\": { \\"continent\\": \\"EU\\", \\"iso_country\\": \\"UA\\", \\"iso_region\\": \\"UA-32\\" }
}
"
`;

exports[`Airports Script output when input is sorted alphabetically should output in the same order 1`] = `
"{
  \\"CPH\\": { \\"continent\\": \\"EU\\", \\"iso_country\\": \\"DK\\", \\"iso_region\\": \\"DK-84\\" },
  \\"IEV\\": { \\"continent\\": \\"EU\\", \\"iso_country\\": \\"UA\\", \\"iso_region\\": \\"UA-32\\" },
  \\"KBP\\": { \\"continent\\": \\"EU\\", \\"iso_country\\": \\"UA\\", \\"iso_region\\": \\"UA-32\\" }
}
"
`;

exports[`Airports Script output when input is unsorted should output sorted 1`] = `
"{
  \\"CPH\\": { \\"continent\\": \\"EU\\", \\"iso_country\\": \\"DK\\", \\"iso_region\\": \\"DK-84\\" },
  \\"IEV\\": { \\"continent\\": \\"EU\\", \\"iso_country\\": \\"UA\\", \\"iso_region\\": \\"UA-32\\" },
  \\"KBP\\": { \\"continent\\": \\"EU\\", \\"iso_country\\": \\"UA\\", \\"iso_region\\": \\"UA-32\\" }
}
"
`;
